src/main/java/com/google/sps/data/RatingEmailTask.java

```diff
@@ -0,0 +1,98 @@
+
+import com.google.sps.data.TimeRange;
+import com.google.cloud.tasks.v2.AppEngineHttpRequest;
+import com.google.cloud.tasks.v2.CloudTasksClient;
+import com.google.cloud.tasks.v2.HttpMethod;
+import com.google.cloud.tasks.v2.QueueName;
+import com.google.cloud.tasks.v2.Task;
+import com.google.common.base.Strings;
+import com.google.protobuf.ByteString;
+import com.google.protobuf.Timestamp;
+import java.util.Calendar;
+import java.nio.charset.Charset;
+import java.io.IOException;
+import java.time.Clock;
+import java.time.Instant;
+
+/**
+ * Task to send a reminder email to the user once the tutoring session ends
+ */
+public class RatingEmailTask {
+    private long scheduledSeconds;
+    private String studentEmail;
+
+    public RatingEmailTask(TimeRange timeslot, String studentEmail) throws IOException {
+        int year = timeslot.getDate().get(Calendar.YEAR);
+        int month = timeslot.getDate().get(Calendar.MONTH);;
+        int day = timeslot.getDate().get(Calendar.DAY_OF_MONTH);;
+        int hour = (int) Math.floor(timeslot.getEnd() / 60);
+        int min = timeslot.getEnd() % 60;
+
+        Calendar scheduledDate = Calendar.getInstance();
+        scheduledDate.set(Calendar.YEAR, year);
+        scheduledDate.set(Calendar.MONTH, month);
+        scheduledDate.set(Calendar.DAY_OF_MONTH, day);
+        scheduledDate.set(Calendar.HOUR_OF_DAY, hour);
+        scheduledDate.set(Calendar.MINUTE, min);
+
+        this.scheduledSeconds = scheduledDate.getTime().getTime();
+        this.studentEmail = studentEmail;
+
+        scheduleTask();
+    }
+
+    public void scheduleTask() throws IOException {
+
+        // Instantiates a client.
+        try (CloudTasksClient client = CloudTasksClient.create()) {
+            
+            String projectId = "icecube-step-2020";
+            String queueName = "rating-email-queue";
+            String location = "us-central1";
+            String payload = this.studentEmail;
+
+            // Construct the fully qualified queue name.
+            String queuePath = QueueName.of(projectId, location, queueName).toString();
+
+            // Construct the task body.
+            Task.Builder taskBuilder =
+                Task.newBuilder()
+                    .setAppEngineHttpRequest(
+                        AppEngineHttpRequest.newBuilder()
+                            .setBody(ByteString.copyFrom(payload, Charset.defaultCharset()))
+                            .setRelativeUri("/rating-email")
+                            .setHttpMethod(HttpMethod.POST)
+                            .build());
+
+            long current = Calendar.getInstance().getTime().getTime();
+
+            // Add the scheduled time to the request.
+            int seconds = (int) (this.scheduledSeconds - current) / 1000;
+            taskBuilder.setScheduleTime(
+                Timestamp.newBuilder()
+                    .setSeconds(Instant.now(Clock.systemUTC()).plusSeconds(seconds).getEpochSecond()));
```

Comment:
In that case, does this just set the seconds to the time of the scheduled session? Couldn't we just use scheduledSessions here?

Reply:
The thing is that the end time in scheduledSessions is the number of milliseconds in total from january 1st 1970. However, what needs to be passed to the task builder is the seconds from the time when the tutoring session is scheduled. This is why I needed to subtract the current time to only get those seconds. If I just used the number of seconds from scheduled sessions it would be the equivalent of scheduling the task for 50 years from now.